### ISSUE TITLE
Replace option

### ISSUE THREAD

**Issue by anotatta** [Genivia/ugrep#179]
Is there any way to do a replacement like in riggrep
--passthru  --replace <REPLACEMENT_TEXT>

Thanks!

**Comment by genivia-inc** [comment]
I had this on the TODO list of features, but refrained from adding this feature to ugrep because `sed` (the stream editor) is typically used for this task and is much better at this than a replacement text. Just run `sed 's/PATTERN/REPLACEMENT/g FILE` to produce the results on standard output, where REPLACEMENT may include backreferences to insert submatches specified in the PATTERN. You can even specify multiple patterns and replacements. One small thing to remember is that PATTERN uses BRE (Basic REgex) syntax by default unless you use option `-E` to specify ERE (the usual) regex syntax. Search for "sed tutorial" online to learn more. Because I use `sed` often in scripts, I can't really say if there is any value to add such a feature to a grep tool.

**Comment by anotatta** [comment]
Both sed's BRE&ERE regex are limited (look ahead/behind, non-greedy, unicode category, etc) and can be problematic with binary files.  grep and sed compliment each other nicely and use similar regex syntax.  With ugrep, one has such a deep and wide ranging regex capability, offering even PCRE.  I haven't seen any "u-sed" yet (There is "sd" but limited in feature).  Please keep the replace feature on the to do list.  The replace feature will make ugrep a grep-sed on steroid.

**Comment by genivia-inc** [comment]
OK!

**Comment by genivia-inc** [comment]
A quick update: I've implemented `--replace` in my dev version and it works well.

As for the replacement string, I'm strongly considering the `FORMAT` strings convention with `%` fields that `--format` uses. This offers a lot of customization of the match replacement output, not only group captures with `%1` to `%9`, `%[NUM]#` and named captures `%[NAME]#`, but also `%m` for the match number, `%n`, `%k` and `%b` for the line number, column number, byte offset (for the specific match), `%o` for the matching string itself, `%[TEXT1|TEXT2|...]g` to output `TEXT`, or `TEXT2` etc depending on the group capture index, and so on (see manual page section FORMAT):
~~~
    --replace=FORMAT
            Replace matching patterns in the output by the specified FORMAT.
            When option -P is used, FORMAT may include `%1' to `%9', `%[NUM]#'
            and `%[NAME]#' to output group captures.  A `%%' outputs `%' and
            `%~' outputs a newline.  See option --format and `man ugrep'
            section FORMAT for details.
~~~

**Comment by anotatta** [comment]
That is delightful news.  Having --replace=FORMAT option would be less confusing to user and leaving room for future enhancement.  

Thank you and may you be well.

**Comment by genivia-inc** [comment]
By the way, `ugrep -P -y --replace=FORMAT` would be the same as `ripgrep --passthru --replace <REPLACEMENT_TEXT>`, but with `%1` for group capture 1 instead of `$1`. IMHO the use of `$` to identify captures is not a good idea, because shells interpret `$` in a `"string"`. As for the use of `-P`, Ripgrep uses Perl matching, like grep `-P`.

I'm thoroughly testing the `--replace` addition and will release an update soon after updating the README, man page and implementing speed improvements to restore more efficient AVX2 optimizations that were lost due to binary portability changes applied to ugrep's build steps made to address requests #103 and #143 in the past.

**Comment by genivia-inc** [comment]
Option `--replace` is available now in ugrep release v3.6.0.
